### PR TITLE
Fix #1058 - Use thread-safe wcsrtombs insead of wcstombs

### LINKED ISSUE #1058: Migrate from `wcstombs()` function since is not thread-safe
`wcstombs` is not thread safe and should be fixed.

This issue is linked to this PR: https://github.com/rizinorg/rizin/pull/1045

image

### PR DESCRIPTION
**Pull request by kholis1**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
Migrated from wcstombs to wcsrtombs for threads safety

...

**Test plan**

...

**Closing issues**
closes #1058 

...

### CODE REVIEW

File: librz/util/str.c [hunk 1]
```diff
@@ -3494,20 +3495,22 @@ RZ_API wchar_t *rz_str_mb_to_wc_l(const char *buf, int len) {
 
 RZ_API char *rz_str_wc_to_mb_l(const wchar_t *buf, int len) {
 	char *res_buf = NULL;
+	mbstate_t mbstate;
```
Comment (wargio) [inline librz/util/str.c]:
```suggestion
	mbstate_t mbstate = { 0 };
```

File: librz/util/str.c [hunk 2]
```diff
@@ -3494,20 +3495,22 @@ RZ_API wchar_t *rz_str_mb_to_wc_l(const char *buf, int len) {
...
+    memset((void*)&mbstate, 0, sizeof(mbstate));
```
Comment (wargio) [inline librz/util/str.c]: you can remove this if you add the change above

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/util/str.c [hunk 3]
```diff
@@ -3494,20 +3495,21 @@ RZ_API wchar_t *rz_str_mb_to_wc_l(const char *buf, int len) {
...
 RZ_API char *rz_str_wc_to_mb_l(const wchar_t *buf, int len) {
 	char *res_buf = NULL;
+	mbstate_t mbstate = { 0 };
 	size_t sz;
 
 	if (!buf || len <= 0) {
 		return NULL;
 	}
-	sz = wcstombs(NULL, buf, 0);
+	sz = wcsrtombs(NULL, &buf, 0, &mbstate);
```
Comment (wargio) [inline librz/util/str.c:3504]: Ok, so according to https://www.cplusplus.com/reference/cwchar/wcsrtombs/ you should use `mbrlen` to initialize `mbstate`. Unfortunately i never used wcsrtombs so i would wait for @karliss to comment on this.

Comment (kholis1) [reply librz/util/str.c:3504]: https://www.gnu.org/software/libc/manual/html_node/Keeping-the-state.html says there are more than one ways to initialize `mbstate` and `memset` is one of them so `mbstate = { 0 };` should work too.

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**